experimental/list_formatter/src/deduplicating_array.rs

```diff
@@ -0,0 +1,111 @@
+use core::{mem, ops::Index, ptr};
+use icu_provider::yoke::*;
+
+#[derive(Debug)]
+#[cfg_attr(
+    feature = "provider_serde",
+    derive(serde::Deserialize, serde::Serialize)
+)]
+// TODO: Make this generic over the 12. Serde currently can't derive
+// if const generics are present: https://github.com/serde-rs/serde/issues/1937
+pub struct DeduplicatingArray<T>([Dedupe<T>; 12]);
```

Comment:
Praise: I think this is a nice useful generic class to introduce.  We may want to move it somewhere more central after this is all checked in and landed.

Reply:
I think until we can make this generic in the length this isn't really reusable. Unless we make this whole file a macro.